src/Query/EagerLoader/Extractor.php

```diff
@@ -0,0 +1,119 @@
+<?php
+namespace Mongolid\Query\EagerLoader;
+
+use MongoDB\BSON\ObjectId;
+use Mongolid\Query\EagerLoader\Exception\EagerLoaderException;
+use Mongolid\Query\EagerLoader\Exception\InvalidModelKeyException;
+
+/**
+ * Responsible for extract ids from the model based on its
+ * relationship. You can use dot notations in order to
+ * extract ids from your embedded models.
+ */
+class Extractor
+{
+    /**
+     * This array will handle all related models extracted
+     * from the models passed by.
+     *
+     * @var array
+     */
+    private $relatedModels;
+
+    public function __construct(array $relatedModels)
+    {
+        $this->relatedModels = $relatedModels;
+    }
+
+    /**
+     * The model received here may be either an array from mongodb
+     * or a cached model as a serialized array.
+     * That's why we always will force an array to be used here.
+     * It will ensure that we are only working with model arrays.
+     */
+    public function extractFrom(array $model): array
+    {
+        foreach ($this->relatedModels as $eagerLoadKey => $loadModel) {
+            $key = $this->extractKeyFrom($loadModel);
+
+            // If the key extract contains '.' notations means that
+            // we want to extract models from embedded relations.
+            if ($this->keyHasDots($key)) {
+                $this->extractFromEmbeddedModel($eagerLoadKey, $model, $key);
+
+                continue;
+            }
+
+            $this->extractFromModel($eagerLoadKey, $model, $key);
+        }
+
+        return $this->relatedModels;
```

Comment:
In order of my suggestion about change this method and avoid `foreachs` nested, maybe you can do this here:
```suggestion
        return $this->relatedModels[$eagerLoadKey];
```